tools/api.py

```diff
@@ -144,41 +145,45 @@ def substitute(self, form):
                                va_start="\n    ".join(self.va_start),
                                va_end="\n    ".join(self.va_end))
 
-    def generate_header(self):
+    def generate_header(self) -> str:
         return self.substitute(header_function)
 
-    def generate_body(self):
+    def generate_body(self) -> str:
         return self.substitute(c_api_impl)
 
 
 class BadParam:
-    def __init__(self, cond, msg):
+    def __init__(self, cond: str, msg: str) -> None:
         self.cond = cond
         self.msg = msg
 
 
 class Parameter:
-    def __init__(self, name, type, optional=False, returns=False):
+    def __init__(self,
+                 name: str,
+                 type: str,
+                 optional: bool = False,
+                 returns: bool = False) -> None:
         self.name = name
         self.type = Type(type)
         self.optional = optional
-        self.cparams = []
+        self.cparams: List[Tuple[str, str]] = []
```

Comment:
Do we need to add types for all other members e.g. `self.name`, `self.optional` or just the required important ones ?